paparazzi/src/main/java/app/cash/paparazzi/internal/apng/ApngReader.kt

```diff
@@ -0,0 +1,265 @@
+    }
+  }
+
+  private fun BufferedSource.parseACTL() {
+    frameCount = readInt()
+  }
+
+  private fun BufferedSource.parseFCTL() {
+    val nextSequence = readInt()
+    frameInfo = FrameInfo(
+      width = readInt(),
+      height = readInt(),
+      offsetX = readInt(),
+      offsetY = readInt(),
+      delayNumerator = readShort(),
+      delayDenominator = readShort()
+    )
+
+    if (sequenceNumber + 1 != nextSequence) {
+      throw IllegalStateException("Out of order sequence, expected: ${sequenceNumber + 1} actual: $nextSequence")
+    }
+    sequenceNumber = nextSequence
+  }
+
+  private fun BufferedSource.parseIDAT() {
+    decompress().decodeToScratchSpace()
+  }
+
+  private fun BufferedSource.parseFDAT() {
+    val nextSequence = readInt()
+    if (sequenceNumber + 1 != nextSequence) {
+      throw IllegalStateException("Out of order sequence, expected: ${sequenceNumber + 1} actual: $nextSequence")
+    }
+    sequenceNumber = nextSequence
+    decompress().decodeToScratchSpace()
+  }
+
+  private fun BufferedSource.decompress(): Buffer {
+    val inflater = Inflater().apply {
+      setInput(readByteArray())
+    }
+
+    val buffer = ByteArray(BUFFER_SIZE)
+    return Buffer().apply {
+      do {
+        val byteCount = inflater.inflate(buffer)
+        write(buffer, 0, byteCount)
+      } while (!inflater.finished() && byteCount != 0)
+    }
+  }
+
+  private fun BufferedSource.decodeToScratchSpace() {
+    val frameWidth = frameInfo?.width ?: width
+    val frameHeight = frameInfo?.height ?: height
+    val frameOffsetX = frameInfo?.offsetX ?: 0
+    val frameOffsetY = frameInfo?.offsetY ?: 0
+
+    for (y in 0 until frameHeight) {
+      val filterByte = readByte()
+      if (filterByte != 0.toByte()) {
+        throw IllegalStateException("Only filter None is supported, $filterByte")
+      }
+
+      for (x in 0 until frameWidth) {
+        val pixelBytes = if (hasAlpha) readByteArray(4) else readByteArray(3)
+        val rgba = Buffer().run {
+          write(pixelBytes)
+          if (!hasAlpha) {
+            writeByte(0)
+          }
+
+          readInt()
+        }
+
+        val argb = rgba.rotateRight(8)
+        val position = ((frameOffsetY + y) * width) + frameOffsetX + x
+        scratchSpace[position] = argb
+      }
+    }
+
+    frameNumber++
+  }
+
+  private fun BufferedSource.readChunk(): Pair<Header, BufferedSource> {
+    val dataLength = readInt()
+    val chunkId = readByteArray(4L)
+    val data = readByteArray(dataLength.toLong())
+    val crc = readInt()
+
+    crcEngine.reset()
+    crcEngine.update(chunkId, 0, 4)
+    if (dataLength > 0) crcEngine.update(data, 0, dataLength)
+
+    if (crcEngine.value.toInt() != crc) {
+      throw IllegalStateException("CRC Mismatch decoding ${chunkId.commonToUtf8String()}, invalid data")
+    }
+
+    return Header.valueOf(chunkId.commonToUtf8String().uppercase()) to Buffer().apply { write(data) }
```

Comment:
Instead of reading chunk to byte array and then converting to string read it directly to a UTF-8 string to avoid allocation. Should just be something like `readUt8String(4)`

Reply:
This ByteArray is unfortunately needed for updating the CRC which takes either the ByteArray or ByteBuffer